src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,673 @@
+			// Free them.
+			return summary_truncation_options(castNode(RawStmt, node)->stmt, state);
+
+		case T_SelectStmt:
+			{
+				SelectStmt *stmt = castNode(SelectStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							select_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				if (stmt->valuesLists != NULL)
+					add_truncation(state,
+							TRUNCATION_VALUES_LISTS,
+							node,
+							select_values_lists_len(stmt->valuesLists));
+
+				break;
+			}
+
+		case T_UpdateStmt:
+			{
+				UpdateStmt *stmt = castNode(UpdateStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_DeleteStmt:
+			{
+				DeleteStmt *stmt = castNode(DeleteStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_CopyStmt:
+			{
+				CopyStmt *stmt = castNode(CopyStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_InsertStmt:
+			{
+				InsertStmt *stmt = castNode(InsertStmt, node);
+
+				if (stmt->cols != NULL)
+					add_truncation(state, TRUNCATION_COLS, node, cols_len(stmt->cols));
+
+				break;
+			}
+
+		case T_IndexStmt:
+			{
+				IndexStmt *stmt = castNode(IndexStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_RuleStmt:
+			{
+				RuleStmt *stmt = castNode(RuleStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_CommonTableExpr:
+			{
+				CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+
+				if (stmt->ctequery != NULL) {
+					PgQueryDeparseResult result = pg_query_deparse(node);
+
+					if (result.error) {
+						state->error = result.error;
+						return false;
+					}
+
+					int32_t length = strlen(result.query);
+
+					add_truncation(state,
+							TRUNCATION_CTE_QUERY,
+							node,
+							length);
+				}
+
+				break;
+			}
+
+		case T_InferClause:
+			{
+				InferClause *stmt = castNode(InferClause, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_OnConflictClause:
+			{
+				OnConflictClause *stmt = castNode(OnConflictClause, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		default:
+			break;
+	}
+
+	if (!pg_query_raw_tree_walker_supports(node))
+		return false;
+
+	state->depth++;
+
+	return raw_expression_tree_walker(node, summary_truncation_options, (void *) state);
```

Comment:
As discussed, this won't correctly behave when there is sibling nodes that the walker gets called for, so we'd either need to pass this to the function as a separate argument (seems challenging to implement), or restore the previous value after having descended down, like this:

```suggestion
	int old_depth = state->depth;
	state->depth++;

	bool result = raw_expression_tree_walker(node, summary_truncation_options, (void *) state);

	// Restore old depth value since the current node (or its parents) may have sibling elements
	state->depth = old_depth;

	return result;
```